xarray/namedarray/core.py

```diff
@@ -0,0 +1,342 @@
+from __future__ import annotations
+
+import copy
+import itertools
+import math
+import typing
+from collections.abc import Hashable, Iterable, Mapping
+
+import numpy as np
+import numpy.typing as npt
+
+from xarray.namedarray.utils import Frozen, _default, is_duck_dask_array
+
+# temporary placeholder for indicating that an array api compliant
+# type. hopefully in the future we can narrow this down more
+T_Array = typing.TypeVar("T_Array", bound=typing.Any)
+
+
+class NamedArray:
+    __slots__ = ("_dims", "_data", "_attrs")
+
+    def __init__(
+        self, dims, data: npt.ArrayLike, attrs: dict[typing.Any, typing.Any] = None
+    ):
+        self._dims = self._parse_dimensions(dims)
+        self._data = data
+        self._attrs = attrs or {}
```

Comment:
Let's type annotate these:
```suggestion
        self._dims: tuple[Hashable, ...] = self._parse_dimensions(dims)
        self._data: DuckArray = data
        self._attrs: dict | None = attrs or {}
```